src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,395 @@
+/**
+ * @file lmnn_function_impl.cpp
+ * @author Manish Kumar
+ *
+ * An implementation of the LMNNFunction class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LMNN_FUNCTION_IMPL_HPP
+#define MLPACK_METHODS_LMNN_FUNCTION_IMPL_HPP
+
+#include "lmnn_function.hpp"
+#include "constraints.hpp"
+
+#include <mlpack/core/math/make_alias.hpp>
+#include <mlpack/core/optimizers/function.hpp>
+
+namespace mlpack {
+namespace lmnn {
+
+template<typename MetricType>
+LMNNFunction<MetricType>::LMNNFunction(const arma::mat& dataset,
+                           const arma::Row<size_t>& labels,
+                           size_t k,
+                           double regularization,
+                           MetricType metric) :
+    dataset(math::MakeAlias(const_cast<arma::mat&>(dataset), false)),
+    labels(math::MakeAlias(const_cast<arma::Row<size_t>&>(labels), false)),
+    k(k),
+    metric(metric),
+    regularization(regularization)
+{
+  // Initialize the initial learning point.
+  initialPoint.eye(dataset.n_rows, dataset.n_rows);
+
+  // Initialize target neighbors & impostors.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+  constraint.Impostors(impostors);
+}
+
+//! Shuffle the dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Shuffle()
+{
+  arma::mat newDataset;
+  arma::Row<size_t> newLabels;
+
+  math::ShuffleData(dataset, labels, newDataset, newLabels);
+
+  math::ClearAlias(dataset);
+  math::ClearAlias(labels);
+
+  dataset = std::move(newDataset);
+  labels = std::move(newLabels);
+
+  // Re-calculate target neighbors as indices changed.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+}
+
+inline void Projection(arma::mat& iterate)
+{
+  // Projection.
+  arma::vec eigval;
+  arma::mat eigvec;
+
+  arma::eig_sym(eigval, eigvec, iterate);
+
+  arma::uvec ind = arma::find(eigval > 0);
```

Comment:
You could check to see if `Projection()` is ever doing anything with something like this:

```
if (ind.n_elem != eigval.n_elem)
  Log::Warn << "Removed " << (eigval.n_elem - ind.n_elem) << " eigenvectors!" << std::endl;
```